includes/class-rp-query.php

```diff
@@ -0,0 +1,372 @@
+				$wp->query_vars[ $key ] = sanitize_text_field( wp_unslash( $_GET[ $var ] ) ); // WPCS: input var ok, CSRF ok.
+			} elseif ( isset( $wp->query_vars[ $var ] ) ) {
+				$wp->query_vars[ $key ] = $wp->query_vars[ $var ];
+			}
+		}
+	}
+
+	/**
+	 * Are we currently on the front page?
+	 *
+	 * @param WP_Query $q Query instance.
+	 * @return bool
+	 */
+	private function is_showing_page_on_front( $q ) {
+		return $q->is_home() && 'page' === get_option( 'show_on_front' );
+	}
+
+	/**
+	 * Is the front page a page we define?
+	 *
+	 * @param int $page_id Page ID.
+	 * @return bool
+	 */
+	private function page_on_front_is( $page_id ) {
+		return absint( get_option( 'page_on_front' ) ) === absint( $page_id );
+	}
+
+	/**
+	 * Hook into pre_get_posts to do the main food query.
+	 *
+	 * @param WP_Query $q Query instance.
+	 */
+	public function pre_get_posts( $q ) {
+		// We only want to affect the main query.
+		if ( ! $q->is_main_query() ) {
+			return;
+		}
+
+		// Fix for endpoints on the homepage.
+		if ( $this->is_showing_page_on_front( $q ) && ! $this->page_on_front_is( $q->get( 'page_id' ) ) ) {
+			$_query = wp_parse_args( $q->query );
+			if ( ! empty( $_query ) && array_intersect( array_keys( $_query ), array_keys( $this->get_query_vars() ) ) ) {
+				$q->is_page     = true;
+				$q->is_home     = false;
+				$q->is_singular = true;
+				$q->set( 'page_id', (int) get_option( 'page_on_front' ) );
+				add_filter( 'redirect_canonical', '__return_false' );
+			}
+		}
+
+		// Fix food feeds.
+		if ( $q->is_feed() && $q->is_post_type_archive( 'food_menu' ) ) {
+			$q->is_comment_feed = false;
+		}
+
+		// Only apply to food categories, the food post archive, the menu page, and food tags.
+		if ( ! $q->is_post_type_archive( 'food_menu' ) && ! $q->is_tax( get_object_taxonomies( 'food_menu' ) ) ) {
+			return;
+		}
+
+		$this->food_query( $q );
+
+		// And remove the pre_get_posts hook.
+		$this->remove_food_query();
+	}
+
+	/**
+	 * Query the foods, applying sorting/ordering etc.
+	 * This applies to the main WordPress loop.
+	 *
+	 * @param WP_Query $q Query instance.
+	 */
+	public function food_query( $q ) {
+		if ( ! is_feed() ) {
+			$ordering  = $this->get_catalog_ordering_args();
+			$q->set( 'orderby', $ordering['orderby'] );
+			$q->set( 'order', $ordering['order'] );
+
+			if ( isset( $ordering['meta_key'] ) ) {
+				$q->set( 'meta_key', $ordering['meta_key'] );
+			}
+		}
+
+		// Query vars that affect posts shown.
+		$q->set( 'meta_query', $this->get_meta_query( $q->get( 'meta_query' ), true ) );
+		$q->set( 'tax_query', $this->get_tax_query( $q->get( 'tax_query' ), true ) );
+		$q->set( 'rp_query', 'food_query' );
+		$q->set( 'post__in', array_unique( (array) apply_filters( 'loop_menu_post_in', array() ) ) );
+
+		// Work out how many foods to query.
+		$q->set( 'posts_per_page', $q->get( 'posts_per_page' ) ? $q->get( 'posts_per_page' ) : apply_filters( 'loop_menu_per_page', 10 ) );
```

Comment:
@rabinstha By default we will have 10 posts per page. Additionally user can play with the filter hook :)